src/PMD/parser/parser.jl

```diff
@@ -0,0 +1,523 @@
+include("states.jl")
+
+mutable struct Parser
+    # io pointer to read from
+    io::IO
+
+    # original pmd path and current line number, used to
+    # compose better error/warning messages
+    path::String
+    line::Int
+
+    # current parser state
+    state::Vector{Any}
+    merge_index::Int
+
+    # whether to display warnings
+    verbose::Bool
+
+    # Study info
+    data_struct::DataStruct
+    relation_mapper::RelationMapper
+    model_template::ModelTemplate
+
+    function Parser(
+        io::IO,
+        path::AbstractString,
+        data_struct::DataStruct,
+        relation_mapper::RelationMapper,
+        model_template::ModelTemplate;
+        verbose::Bool = false,
+    )
+        return new(
+            io,
+            path,
+            0,  # line
+            [], # state
+            1,  # merge_index
+            verbose,
+            data_struct,
+            relation_mapper,
+            model_template,
+        )
+    end
+end
+
+function _get_state(parser::Parser)
+    if isempty(parser.state)
+        return PMD_IDLE()
+    else
+        return parser.state[end]
+    end
+end
+
+function _push_state!(parser::Parser, state::ParserState)
+    push!(parser.state, state)
+
+    return nothing
+end
+
+function _pop_state!(parser::Parser)
+    if isempty(parser.state)
+        error("Empty parser state stack")
+    else
+        return pop!(parser.state)
+    end
+end
+
+function _syntax_error(
+    parser::Parser,
+    msg::AbstractString,
+)
+    return error("Syntax Error in '$(parser.path):$(parser.line)': $msg")
+end
+
+function _syntax_warning(
+    parser::Parser,
+    msg::AbstractString,
+)
+    if parser.verbose
+        @warn("In '$(parser.path):$(parser.line)': $msg")
+    end
+end
+
+function _cache_merge!(parser::Parser, collection::String, model_name::String)
+    name = parser.model_template.inv[model_name]
+
+    attribute = Attribute(name, false, DataType, 0, "")
+
+    parser.data_struct[collection]["_MERGE_$(parser.merge_index)"] = attribute
+
+    parser.merge_index += 1
+
+    return nothing
+end
+
+function _apply_tag!(
+    parser::Parser,
+    tag::String,
+    state::S,
+) where {S <: Union{PMD_DEF_MODEL, PMD_DEF_CLASS}}
+    if tag == "@id"
+        _syntax_warning(
+            parser,
+            "Unhandled '$tag' tag found within '$(state.collection)' definition",
+        )
+    elseif tag == "@hourly_dense"
+        _syntax_warning(
+            parser,
+            "Unhandled '$tag' tag found within '$(state.collection)' definition",
+        )
+    else
+        _syntax_error(parser, "Unknown tag '$tag' within '$(state.collection)'")
+    end
+end
+
+function parse end
```

Comment:
You should either put a docstring here or delete this empty definition

Reply:
We are adding docstrings

Reply:
Done